Server/ReasnAPI/ReasnAPI/Services/EventService.cs

```diff
@@ -1,16 +1,255 @@
-         * get by ID
-         * get list by filter
-         * get all
-         */
-    }
-}
+﻿using Microsoft.EntityFrameworkCore;
+using ReasnAPI.Exceptions;
+using ReasnAPI.Models.Database;
+using ReasnAPI.Models.DTOs;
+using System.Linq.Expressions;
+using System.Transactions;
+using ReasnAPI.Models.Enums;
+using ReasnAPI.Models.Mappers;
+using ReasnAPI.Exceptions;
+using System.Text.RegularExpressions;
+
+namespace ReasnAPI.Services;
+public class EventService(ReasnContext context)
+{
+    public EventDto CreateEvent(EventDto eventDto)
+    {
+        using (var scope = new TransactionScope())
+        {
+            
+            var createdTime = DateTime.UtcNow;
+            var newEvent = eventDto.ToEntity();
+            newEvent.CreatedAt = createdTime;
+            newEvent.UpdatedAt = createdTime;
+            newEvent.Slug = CreateSlug(eventDto, createdTime);
+           
+
+            context.Events.Add(newEvent);
+            context.SaveChanges();
+
+            var eventId = newEvent.Id;
+
+            if (eventDto.Tags is not null && eventDto.Tags.Any())
+            {
+                var newTags = eventDto.Tags
+                    .Where(t => !context.Tags.Any(x => x.Name == t.Name))
+                    .Select(t => new Tag { Name = t.Name })
+                    .ToList();
+
+                context.Tags.AddRange(newTags);
+                newEvent.Tags = newTags;
+            }
+
+            if (eventDto.Parameters is not null && eventDto.Parameters.Any())
+            {
+                var newParameters = eventDto.Parameters
+                    .Where(p => !context.Parameters.Any(x => x.Key == p.Key && x.Value == p.Value))
+                    .Select(p => new Parameter { Key = p.Key, Value = p.Value })
+                    .ToList();
+
+                context.Parameters.AddRange(newParameters);
+
+                newEvent.Parameters = newParameters;
+            }
+
+            context.SaveChanges();
+            scope.Complete();
+        }
+
+        return eventDto;
+    }
+
+    public EventDto UpdateEvent(int eventId, EventDto eventDto)
+    {
+        using (var scope = new TransactionScope())
+        {
+            var eventToUpdate = context.Events.Include(e => e.Tags).Include(e => e.Parameters).FirstOrDefault(e => e.Id == eventId);
+
+            if (eventToUpdate is null)
+            {
+                throw new NotFoundException("Event not found");
+            }
+
+            eventToUpdate.Name = eventDto.Name;
+            eventToUpdate.AddressId = eventDto.AddressId;
+            eventToUpdate.Description = eventDto.Description;
+            eventToUpdate.OrganizerId = eventDto.OrganizerId;
+            eventToUpdate.StartAt = eventDto.StartAt;
+            eventToUpdate.EndAt = eventDto.EndAt;
+            eventToUpdate.UpdatedAt = DateTime.Now;
+            eventToUpdate.Status = eventDto.Status;
+
+            context.Events.Update(eventToUpdate);
+            context.SaveChanges();

```

Comment:
I think we can even call it at the end, so when we attach tags to eventToUpdate and parameters then call update and savechanges at once for that, wdyt?

Reply:
good idea